ogre/src/OgreHeightmap.cc

```diff
@@ -0,0 +1,3304 @@
+                     Ogre::StringStream &_outStream);
+
+      protected: virtual void generateVpDynamicShadows(
+                     const SM2Profile *_prof, const Ogre::Terrain *_terrain,
+                     TechniqueType _tt,
+                     Ogre::StringStream &_outStream);
+
+      protected: virtual void generateFpHeader(const SM2Profile *_prof,
+                     const Ogre::Terrain *_terrain,
+                     TechniqueType tt,
+                     Ogre::StringStream &_outStream);
+
+      protected: virtual void generateFpLayer(const SM2Profile *_prof,
+                     const Ogre::Terrain *_terrain, TechniqueType tt,
+                     Ogre::uint _layer,
+                     Ogre::StringStream &_outStream);
+
+      protected: virtual void generateFpFooter(const SM2Profile *_prof,
+                     const Ogre::Terrain *_terrain,
+                     TechniqueType tt,
+                     Ogre::StringStream &_outStream);
+
+      protected: virtual void generateFpDynamicShadowsParams(
+                     Ogre::uint *_texCoord, Ogre::uint *_sampler,
+                     const SM2Profile *_prof, const Ogre::Terrain *_terrain,
+                     TechniqueType _tt,
+                     Ogre::StringStream &_outStream);
+
+      protected: virtual void generateFpDynamicShadowsHelpers(
+                     const SM2Profile *_prof,
+                     const Ogre::Terrain *_terrain,
+                     TechniqueType tt,
+                     Ogre::StringStream &_outStream);
+
+      protected: void generateFpDynamicShadows(const SM2Profile *_prof,
+                     const Ogre::Terrain *_terrain, TechniqueType _tt,
+                     Ogre::StringStream &_outStream);
+
+      protected: virtual void generateFragmentProgramSource(
+                     const SM2Profile *_prof,
+                     const Ogre::Terrain *_terrain,
+                     TechniqueType _tt,
+                     Ogre::StringStream &_outStream);
+
+      protected: virtual void updateVpParams(const SM2Profile *_prof,
+                     const Ogre::Terrain *_terrain, TechniqueType _tt,
+                     const Ogre::GpuProgramParametersSharedPtr &_params);
+
+      private: Ogre::String GetChannel(Ogre::uint _idx);
+
+      /// \brief Capabilities
+      private: std::string glslVersion = "130";
+      private: std::string vpInStr = "in";
+      private: std::string vpOutStr = "out";
+      private: std::string fpInStr = "in";
+      private: std::string fpOutStr = "out";
+      private: std::string textureStr = "texture";
+    };
+
+    // Needed to allow access from ShaderHelperGLSL to protected members
+    // of SM2Profile.
+    friend ShaderHelperGLSL;
+
+    /// Keeping the CG shader for reference.
+    /// \brief Utility class to help with generating shaders for Cg / HLSL.
+    /// Original implementation from Ogre that generates Cg shaders
+    protected: class ShaderHelperCg :
```

Comment:
we can actually remove all `ShaderHelperCg` code and references to ShaderHelperCg. It was disabled and not being maintained in gazebo classic, and ign-rendering is not loading the Cg plugin.

Reply:
Removing code feels good! 915fdd495d141b01a44c8259947b1f2b2cf95d5f